# -*- coding: utf-8 -*- 
#Check that the genes in each bin (outputted from check_gene_in_bin_dump_all.py) have orthologs in E. scolopes/O. bimaculoides/P. maximus and output them.
#Interaction will be written to outfile if both genes in the interaction have orthologs in the other species. 
#If only one gene has an ortholog, the interaction will be written to the outfile with "NA" in the column for the gene without an ortholog. 
#If neither gene has an ortholog, the interaction will not be written to the outfile.
#=============================================================================
import argparse
import sys
from collections import defaultdict
import os
import random
import collections
#==============================================================================
#Command line options==========================================================
#==============================================================================
parser = argparse.ArgumentParser()


parser.add_argument("bh_orthos", type=str,
                    help="File of reciprocol best hit orthologs, with Euprymna scolopes or Octopus bimaculoides (depending on the interactions) in the first column")

parser.add_argument("int_file", type=str,
                    help="File of chromosomes, bin IDs, genes and interaction frequencies generated by check_genes_in_bin_dump(_all).py script")


if len(sys.argv)==1:
	parser.print_help()
	sys.exit(1)
args = parser.parse_args()

#==============================================================================
#==============================================================================

def main():

    species1 = args.bh_orthos.split("/")[-1].split("gene")[0] #Name of species 1
    species2 = args.bh_orthos.split("/")[-1].split("gene")[1].split(".txt")[0] #Name of species 2

    spp1_ortho_dict = {} #Dictionary of euprymna and other species orthologs
    with open(args.bh_orthos, "r") as orthos:
        for line in orthos:
            col = line.strip().split("\t") #split already here, so no need to split for individual columns
            spp1_g = col[0]
            spp2_g = col[1] #other species ortholog
            spp1_ortho_dict[spp1_g] = spp2_g
    
    print("Number of reciprocol best hit orthologs for", species1, "and", species2, "=", len(spp1_ortho_dict))

    ortho_int_list = []
    with open(args.int_file, "r") as gene_ints:
        outname = args.int_file.split(".txt")[0] + "_" + species2 + "orthos.txt"
        with open(outname, "w") as outfile:
            outfile.write("Ortholog_pair_genes_bin1\tOrtholog_pair_genes_bin2\tInteraction_frequency\n")
            for line in gene_ints:
                line = line.rstrip()
                if not line.startswith("chromosome"):
                    genes_bin1 = line.split("\t")[2].split(",")
                    genes_bin2 = line.split("\t")[4].split(",")
                    int_freq = line.split("\t")[5]
                    orthos_bin1 = []
                    orthos_bin2 = []
                    counter1 = 0
                    counter2 = 0
                    counter3 = 0
                    for i in genes_bin1:
                        if i in spp1_ortho_dict:
                            orthos_bin1.append(i)
                            counter1 += 1 
                            ortholog1 = spp1_ortho_dict[i]
                            outfile.write(f"{i};{ortholog1},")
                    if counter1 > 0:
                        last_ortho1 = orthos_bin1[-1]
                        for i in genes_bin1:
                            if i in spp1_ortho_dict and i == last_ortho1: #This bit loops through the list again because you had to close it for the list previously.
                                outfile.write("\t") #This checks if the gene has an ortholog and is the last one in the genes_in_bin list for one of the interactions on that line. If it is, tab to start next column.
                    elif counter1 == 0: #This bit writes orthologs to line if only interaction 2 has orthologs but not if neither do.
                        for x in genes_bin2:
                            if x in spp1_ortho_dict:
                                counter3 =+ 1 
                        if counter3 > 0:
                            outfile.write(f"\tNA\t")
                    for x in genes_bin2: #Loop through the list again because you had to close it for the counter previously.
                        if x in spp1_ortho_dict:
                            counter2 =+ 1 
                            orthos_bin2.append(x)
                            ortholog2 = spp1_ortho_dict[x]
                            outfile.write(f"{x};{ortholog2},")
                    if counter2 > 0: 
                        last_ortho2 = orthos_bin2[-1]
                        for x in genes_bin2: #Loop through the list again because you had to close it for the list previously.
                            if x in spp1_ortho_dict and x == last_ortho2: #This checks if the gene has an ortholog and is the last one in the genes_in_bin list for one of the interactions on that line. If it is, tab to start next column and write interaction frequency.
                                outfile.write(f"\t{int_freq}\n")
                    elif counter2 == 0 and counter1 > 0: #Write orthologs to line if only interaction 2 has orthologs but not if neither do
                        outfile.write(f"NA\t{int_freq}\n")
                    if counter1 > 0 and counter2 > 0:
                        ortho_int_list.append(int_freq)

    print("Number of", species1, "interactions with at least one ortholog in", species2, "(NOTE there are duplicates in this file) =", len(ortho_int_list))
    print("Output written to:", outname)
    
if __name__ == "__main__":
    main()
