# -*- coding: utf-8 -*- 
#Check all pairs of genes in O. bimaculoides and E. scolopes and compute their distances.
#Input files with orthologous gene pairs must be formatted for one gene pair per row.
#=============================================================================
#=============================================================================
import argparse
import sys
from collections import defaultdict
import os
import random
import collections
#==============================================================================
#Command line options==========================================================
#==============================================================================
parser = argparse.ArgumentParser()

parser.add_argument("spp_bed_files", type=str ,nargs=2, help="Bed files of chromosome | start | end | gene") #Do this if using multiple files that are in the same file format


parser.add_argument("int_ortho_file", type=str,
                    help="File of orthologous genes and interaction frequencies generated by check_ortho_dump.py script that matches the species in the first BED file. This must be formatted for one gene pair per row.")

if len(sys.argv)==1:
	parser.print_help()
	sys.exit(1)
args = parser.parse_args()
#==============================================================================
#==============================================================================

def main():

    gene_loc_spp1 = {} #Dictionary of gene location for species 1
    gene_loc_spp2 = {} #Dictionary of gene location for species 2
    counter_spp = 0
    for i in args.spp_bed_files:
        counter_spp += 1
        print("Species bed file opened =", i) #This shows what files are being opened
        with open(i, "r") as spp_bed_file:
            for line in spp_bed_file:
                line = line.rstrip()
                chrom = line.split("\t")[0]
                start = line.split("\t")[1]
                end = line.split("\t")[2]
                gene = line.split("\t")[3]
                loc =  chrom + ":" + start + ":" + end
                if counter_spp == 1:
                    gene_loc_spp1[gene] = loc
                    species1 = i.split("/")[-1].split(".")[0] #Name of species 1
                elif counter_spp == 2:
                    gene_loc_spp2[gene] = loc
                    species2 = i.split("/")[-1].split(".")[0] #Name of species 2

    print("NOTE: Third file must have", species1, "genes before the semi colon and", species2, "genes in after the semi colon in first column in order for this script to work.")
    

    counter1 = 0
    with open(args.int_ortho_file, "r") as ortho_ints:
        outname1 = args.int_ortho_file.split(".dumped")[0] + "_" + species1 + "_genom_dist_orthos.txt"
        with open(outname1, "w") as outfile1:
            outfile1.write("Orth_pair_based_on_ " + species1 + "_interaction_matrix")
            outfile1.write("\t")
            outfile1.write("Genomic_distance_" + species1 + "_bp")
            outfile1.write("\n")
            for line in ortho_ints:
                if not line.startswith("Interact"):
                    line = line.rstrip()
                    gene_pair = line.split("\t")[0]
                    spp1_gene1 = gene_pair.split(",")[0].split(";")[0]
                    spp1_gene2 = gene_pair.split(",")[1].split(";")[0]
                    if spp1_gene1 in gene_loc_spp1:
                            loc1 = gene_loc_spp1[spp1_gene1]
                            chrom1 = loc1.split(":")[0]
                            start1 = loc1.split(":")[1]
                            end1 = loc1.split(":")[2]
                            if not chrom1.startswith("scaffold") and not chrom1.startswith("Sc"): #No need to change this to add pecten because the bed file doesn't have scaffolds in.
                                if spp1_gene2 in gene_loc_spp1:
                                    loc2 = gene_loc_spp1[spp1_gene2]
                                    chrom2 = loc2.split(":")[0]
                                    start2 = loc2.split(":")[1]
                                    end2 = loc2.split(":")[2]
                                    if chrom1 == chrom2:
                                        if int(start2) > int(start1) and int(end2) > int(end1): #If first gene comes first (and is not the same gene
                                            counter1 += 1
                                            outfile1.write(gene_pair)
                                            outfile1.write("\t")
                                            genom_dist = int(start2) - int(end1) 
                                            if genom_dist > 0:
                                                outfile1.write(str(genom_dist))
                                            if genom_dist <= 0: #If the octopus/pecten genes are overlapping, give a value of zero.
                                                outfile1.write("0")
                                            outfile1.write("\n")
                                        elif int(start2) < int(start1) and int(end2) < int(end1): #If second gene comes first (and is not the same gene)
                                            counter1 += 1
                                            outfile1.write(gene_pair)
                                            outfile1.write("\t")
                                            genom_dist = int(start1) - int(end2)
                                            if genom_dist > 0: #If the octopus/pecten genes are overlapping, give a value of zero.
                                                outfile1.write(str(genom_dist))
                                            if genom_dist <= 0:
                                                outfile1.write("0")
                                            outfile1.write("\n")

     #Open the first input file again and write out the genomic distance in the other species.                                   
    counter2 = 0
    with open(args.int_ortho_file, "r") as ortho_ints:
        outname2 = args.int_ortho_file.split(".dumped")[0] + "_" + species2 + "_genom_dist_orthos.txt"
        with open(outname2, "w") as outfile2:
            outfile2.write("Orth_pair_based_on_" + species1 + "_interaction_matrix")
            outfile2.write("\t")
            outfile2.write("Genomic_distance_" + species2 + "_bp")
            outfile2.write("\n")
            for line in ortho_ints:
                if not line.startswith("Interact"):
                    line = line.rstrip()
                    gene_pair = line.split("\t")[0]
                    spp2_gene1 = gene_pair.split(",")[0].split(";")[1]
                    spp2_gene2 = gene_pair.split(",")[1].split(";")[1]
                    if spp2_gene1 in gene_loc_spp2:
                        loc1 = gene_loc_spp2[spp2_gene1]
                        chrom1 = loc1.split(":")[0]
                        start1 = loc1.split(":")[1]
                        end1 = loc1.split(":")[2]
                        if not chrom1.startswith("scaffold") and not chrom1.startswith("Sc") : #No need to change this to add pecten because the bed file doesn't have scaffolds in.
                            if spp2_gene2 in gene_loc_spp2:
                                loc2 = gene_loc_spp2[spp2_gene2]
                                chrom2 = loc2.split(":")[0]
                                start2 = loc2.split(":")[1]
                                end2 = loc2.split(":")[2]
                                if chrom1 == chrom2:
                                    if int(start2) > int(start1) and int(end2) > int(end1): #If first gene comes first (and is not the same gene
                                        counter2 += 1
                                        outfile2.write(gene_pair)
                                        outfile2.write("\t")
                                        genom_dist = int(start2) - int(end1) 
                                        if genom_dist > 0:
                                            outfile2.write(str(genom_dist))
                                        if genom_dist <= 0: #If the octopus/pecten genes are overlapping, give a value of zero.
                                            outfile2.write("0")
                                        outfile2.write("\n")
                                    elif int(start2) < int(start1) and int(end2) < int(end1): #If second gene comes first (and is not the same gene)
                                        counter2 += 1
                                        outfile2.write(gene_pair)
                                        outfile2.write("\t")
                                        genom_dist = int(start1) - int(end2)
                                        if genom_dist > 0: #If the octopus/pecten genes are overlapping, give a value of zero.
                                            outfile2.write(str(genom_dist))
                                        if genom_dist <= 0:
                                            outfile2.write("0")
                                        outfile2.write("\n")
                    

    print("Number of intrachromosomal interactions with orthologs in", species1, "and their genomic distances written to output file = ", counter1)
    print("Number of intrachromosomal interactions with orthologs in", species2, "and their genomic distances written to output file = ", counter2)
    print("Note should have 'swapped bin duplicates' removed if you use orthos_form files but there are duplicates because bin pairs have mulitple interaction frequencies but the same genomic distance.")

if __name__ == "__main__":
    main()
